### PR TITLE
Add support for Bazel 6 & 7

### PR DESCRIPTION
**Pull request by fzakaria**
Previous PR #2006 added support for Bazel 7 but some of the changes in Bazel 7 are not backwards compatible with 6.

This is a problem for some of our consumers that embed our project in other Bazel workspaces.

Specifically, the name of the runfiles directory was changed from the workspace name to be '_main'.

The solution adopted here is to use @python_rules and programmatically determine the runfiles directory relative to the lit.cfg.py file.

### CODE REVIEW

**Comment by fzakaria** [comment]
@sjain-stanford do you want try this?
I'm still not sure what's the more idiomatic way -- this approach or doing it in the BUILD file itself.

File: build_tools/utils.bzl [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+def is_bzlmod_enabled():
+    """Determine whether bzlmod mode is enabled."""
+
+    # If bzlmod is enabled, then `str(Label(...))` returns a canonical label,
+    # these start with `@@`.
+    return str(Label("//:invalid")).startswith("@@")
+
+def workspace_name():
+    """Return the name of the workspace."""
+    return "_main" if is_bzlmod_enabled() else "stablehlo"
```
Comment (fzakaria) [inline build_tools/utils.bzl:12]: This isn't used but I'd like to commit it to retain the use if needed.

**Comment by fzakaria** [comment]
Let's revisit when you integrate to confirm since the purpose of the change is to support your integrate.
I don't expect the PR to go stale too much.

**Comment by sjain-stanford** [comment]
@fzakaria do you mind rebasing this branch on latest HEAD? I'm seeing some merge conflicts that I'm not too familiar resolving:
```
Auto-merging MODULE.bazel.lock                                                                                                                                                                             
CONFLICT (content): Merge conflict in MODULE.bazel.lock                                                                                                                                                    
Auto-merging stablehlo/conversions/linalg/tests/BUILD.bazel                                                                                                                                                
CONFLICT (content): Merge conflict in stablehlo/conversions/linalg/tests/BUILD.bazel                                                                                                                       
Auto-merging stablehlo/conversions/tosa/tests/BUILD.bazel                                                                                                                                                  
CONFLICT (content): Merge conflict in stablehlo/conversions/tosa/tests/BUILD.bazel                                                                                                                         
Auto-merging stablehlo/testdata/BUILD.bazel                                                                                                                                                                
CONFLICT (content): Merge conflict in stablehlo/testdata/BUILD.bazel                                                                                                                                       
Auto-merging stablehlo/tests/BUILD.bazel                                                                                                                                                                   
CONFLICT (content): Merge conflict in stablehlo/tests/BUILD.bazel                                                                                                                                          
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by fzakaria** [comment]
@sjain-stanford updated.

File: stablehlo/conversions/tosa/tests/BUILD.bazel [hunk 2]
```diff
@@ -25,9 +25,15 @@ expand_template(
...
-        "\"@STABLEHLO_TOOLS_DIR@\"": "os.path.join(os.environ['TEST_SRCDIR'], '_main')",
-        "@LIT_SITE_CFG_IN_HEADER@": "# Autogenerated, do not edit.",
+        "\"@STABLEHLO_SOURCE_DIR@\"": "RUNFILES_DIR",
+        "\"@STABLEHLO_TOOLS_DIR@\"": "RUNFILES_DIR",
+        "@LIT_SITE_CFG_IN_HEADER@": '''# Autogenerated, do not edit.
+from python.runfiles import Runfiles
+from pathlib import Path
+
+r = Runfiles.Create()
+LITE_CFG_PY = Path(r.Rlocation("stablehlo/stablehlo/conversions/tosa/tests/lite.cfg.py"))
```
Comment (sjain-stanford) [inline stablehlo/conversions/tosa/tests/BUILD.bazel:35]: same

File: stablehlo/testdata/BUILD.bazel [hunk 3]
```diff
@@ -25,9 +25,15 @@ expand_template(
...
-        "\"@STABLEHLO_TOOLS_DIR@\"": "os.path.join(os.environ['TEST_SRCDIR'], '_main')",
-        "@LIT_SITE_CFG_IN_HEADER@": "# Autogenerated, do not edit.",
+        "\"@STABLEHLO_SOURCE_DIR@\"": "RUNFILES_DIR",
+        "\"@STABLEHLO_TOOLS_DIR@\"": "RUNFILES_DIR",
+        "@LIT_SITE_CFG_IN_HEADER@": '''# Autogenerated, do not edit.
+from python.runfiles import Runfiles
+from pathlib import Path
+
+r = Runfiles.Create()
+LITE_CFG_PY = Path(r.Rlocation("stablehlo/stablehlo/testdata/lite.cfg.py"))
```
Comment (sjain-stanford) [inline stablehlo/testdata/BUILD.bazel:35]: same

**Comment by sjain-stanford** [comment]
@fzakaria confirming the patch works for downstream stablehlo integrations 🎉 

```
$ bazel test @stablehlo//...

INFO: Analyzed 4183 targets (25 packages loaded, 9891 targets configured).                                                                                                                                                                                                     
INFO: Found 152 targets and 4031 test targets...                                                                                                                                                                                                                               
[10,255 / 10,277] 78 / 4031 tests; checking cached actions; last test: @stablehlo//stablehlo/t...
.
.
.                                                                                                                                                                        
@stablehlo//stablehlo/tests:vhlo/vhlo_to_version_downgrade_invalid.0_9_0.mlir.test PASSED in 1.1s                                                                                                                                                                              
@stablehlo//stablehlo/tests:vhlo/vhlo_to_version_invalid.mlir.test       PASSED in 1.5s                                                                                                                                                                                        
                                                                                                                                                                                                                                                                               
Executed 4031 out of 4031 tests: 4031 tests pass.                     
```

> I'm still not sure what's the more idiomatic way -- this approach or doing it in the BUILD file itself.

Since the llvm lit config setup is pretty much munging paths anyways, I don't have a strong preference one way or another (doing it in lit.cfg.py vs BUILD). Please go with your best judgment here.

Once this lands, I will go ahead and bump to stablehlo@HEAD in `mlir-tcp` (https://github.com/cruise-automation/mlir-tcp/pull/48).

**Review by sjain-stanford — APPROVED** [review]
Thanks for helping restore the builds in our downstream repos.

File: stablehlo/conversions/linalg/tests/BUILD.bazel [hunk 4]
```diff
@@ -25,9 +25,15 @@ expand_template(
...
-        "\"@STABLEHLO_TOOLS_DIR@\"": "os.path.join(os.environ['TEST_SRCDIR'], '_main')",
-        "@LIT_SITE_CFG_IN_HEADER@": "# Autogenerated, do not edit.",
+        "\"@STABLEHLO_SOURCE_DIR@\"": "RUNFILES_DIR",
+        "\"@STABLEHLO_TOOLS_DIR@\"": "RUNFILES_DIR",
+        "@LIT_SITE_CFG_IN_HEADER@": '''# Autogenerated, do not edit.
+from python.runfiles import Runfiles
+from pathlib import Path
+
+r = Runfiles.Create()
+LITE_CFG_PY = Path(r.Rlocation("stablehlo/stablehlo/conversions/linalg/tests/lite.cfg.py"))
```
[the remark this answers is not in the document]
Comment (fzakaria) [reply stablehlo/conversions/linalg/tests/BUILD.bazel:35]: Let me fix them to be accurate but hes; its fine since it just goes up X parents.

**Review by sdasgup3 — APPROVED** [review]

**Pull request merged**